Add integration with redshift

earmenda opened the pull request:
Closes #254 

### Code Changes

* [x] Add new `redshift` connector to pypi
* [x] Add function `get_redshift_collections_and_fields` for generating db collections
* [x] Add integration tests for Redshift
* [x] Refactor generate-dataset functions to avoid so much repeated code
* [x] Added github secret for Redshift cluster `fidesctl` user. 

### Steps to Confirm

* [x] Manual testing included below
* [x] Integration tests test new Redshift integration

### Pre-Merge Checklist

* [x] All CI Pipelines Succeeded
* [x] Documentation Updated

### Description Of Changes

For integration tests I used an existing test cluster in our AWS account. I created a new database `fidesctl_test` and a user which only has permissions to that database. 
```
CREATE DATABASE FIDESCTL_TEST;x
CREATE SCHEMA public;
```

Within db:
```
CREATE USER fidesctl password '<REDACTED>';

ALTER DEFAULT PRIVILEGES FOR USER fidesctl GRANT ALL ON TABLES to fidesctl;
GRANT ALL ON SCHEMA public to fidesctl;
```

Manually tested `generate-dataset ` and we get the expected output. 

```
/fides/fidesctl# fidesctl generate-dataset "redshift+psycopg2://fidesctl:$REDSHIFT_FIDESCTL_PASSWORD@redshift-cluster-1.cohs2e5eq2e4.us-east-1.redshift.amazonaws.com:5439/fidesctl_test" dataset.yml
```

```
dataset:
- fides_key: public
  organization_fides_key: default_organization
  name: public
  description: 'Fides Generated Description for Schema: public'
  meta: null
  data_categories: []
  data_qualifier: aggregated.anonymized.unlinked_pseudonymized.pseudonymized.identified
  collections:
  - name: public.login
    description: 'Fides Generated Description for Table: public.login'
    data_categories: []
    data_qualifier: aggregated.anonymized.unlinked_pseudonymized.pseudonymized.identified
    fields:
    - name: customer_id
      description: 'Fides Generated Description for Column: customer_id'
      data_categories: []
      data_qualifier: aggregated.anonymized.unlinked_pseudonymized.pseudonymized.identified
      fields: null
    - name: id
      description: 'Fides Generated Description for Column: id'
      data_categories: []
      data_qualifier: aggregated.anonymized.unlinked_pseudonymized.pseudonymized.identified
      fields: null
    - name: time
      description: 'Fides Generated Description for Column: time'
      data_categories: []
      data_qualifier: aggregated.anonymized.unlinked_pseudonymized.pseudonymized.identified
      fields: null
  - name: public.visit
    description: 'Fides Generated Description for Table: public.visit'
    data_categories: []
    data_qualifier: aggregated.anonymized.unlinked_pseudonymized.pseudonymized.identified
    fields:
    - name: email
      description: 'Fides Generated Description for Column: email'
      data_categories: []
      data_qualifier: aggregated.anonymized.unlinked_pseudonymized.pseudonymized.identified
      fields: null
    - name: last_visit
      description: 'Fides Generated Description for Column: last_visit'
      data_categories: []
      data_qualifier: aggregated.anonymized.unlinked_pseudonymized.pseudonymized.identified
      fields: null
```

earmenda commented:
Before merging I wanted to integrate some refactoring @brentonmallen1 suggested. 

Also @SteveDMurphy, if you look at the new output of `generate-dataset` about it includes a null child fields field `fields: null`. Do you think we should exclude this from the output?

SteveDMurphy commented:
> Before merging I wanted to integrate some refactoring @brentonmallen1 suggested.
> 
> Also @SteveDMurphy, if you look at the new output of `generate-dataset` about it includes a null child fields field `fields: null`. Do you think we should exclude this from the output?

@earmenda nice find! I would agree with removing the null child field, it doesn't add any value otherwise and would make for unnecessarily larger yml to wrangle... let me know if you want to pair on it or if we want to create a follow on issue for that work (we didn't create any of the other follow on items either, e.g. connector for mongo, `generate-dataset`, etc. but I don't think what you are suggesting should have to wait)

earmenda commented:
> > Before merging I wanted to integrate some refactoring @brentonmallen1 suggested.
> > Also @SteveDMurphy, if you look at the new output of `generate-dataset` about it includes a null child fields field `fields: null`. Do you think we should exclude this from the output?
> 
> @earmenda nice find! I would agree with removing the null child field, it doesn't add any value otherwise and would make for unnecessarily larger yml to wrangle... let me know if you want to pair on it or if we want to create a follow on issue for that work (we didn't create any of the other follow on items either, e.g. connector for mongo, `generate-dataset`, etc. but I don't think what you are suggesting should have to wait)

@SteveDMurphy I was torn on not including null fields with generate-dataset but I can also imagine how we might want that if a field doesn't have a default value? Simplest solution would be to make the default for `fields` be and empty array, similar to data-categories  what do you think?

earmenda commented on fidesctl/src/fidesctl/core/generate_dataset.py:
```diff
@@ -11,95 +11,37 @@
-        if schema not in schema_exclusion_list:
-            db_tables[schema] = {}
-            for table in inspector.get_table_names(schema=schema):
-                db_tables[schema][f"{schema}.{table}"] = [
```
hi @brentonmallen1 @ThomasLaPiana, I(more like Brenton) refactored this logic so there's not so much repeated code. As part of that I changed the table name in our dataset to no longer include the schema -  `{schema}.{table}` to `{table}`. I don't think we could come up with a good reason to keep this, and it simplifies the logic a bit. If you feel that we should keep this let's discuss that here, otherwise this PR is ok to merge.

ThomasLaPiana replied:
that makes total sense to me, good call here @earmenda / @brentonmallen1

ThomasLaPiana replied:
upon further reading, this code cleanup is top-tier, looks way better now and also simplifies the testing. good work here team!

SteveDMurphy commented:
> > > Before merging I wanted to integrate some refactoring @brentonmallen1 suggested.
> > > Also @SteveDMurphy, if you look at the new output of `generate-dataset` about it includes a null child fields field `fields: null`. Do you think we should exclude this from the output?
> > 
> > 
> > @earmenda nice find! I would agree with removing the null child field, it doesn't add any value otherwise and would make for unnecessarily larger yml to wrangle... let me know if you want to pair on it or if we want to create a follow on issue for that work (we didn't create any of the other follow on items either, e.g. connector for mongo, `generate-dataset`, etc. but I don't think what you are suggesting should have to wait)
> 
> @SteveDMurphy I was torn on not including null fields with generate-dataset but I can also imagine how we might want that if a field doesn't have a default value? Simplest solution would be to make the default for `fields` be and empty array, similar to data-categories what do you think?

@earmenda I see what you mean, I like the consistent behavior as well and if that is easiest to keep moving for now I am on board. Part of me still likes not having anything there if possible but I'm not sure how easy that would be to implement or (worse) if it would make it more mysterious to a user. Also we should check back with @pattisdr as well to see if there are any implications with how `fidesops` will interact with nested fields here (maybe even as a trailing issue) - thanks again for catching and working through that!

SteveDMurphy commented on fidesctl/dataset.yml:
```diff
@@ -0,0 +1,44 @@
+dataset:
```
I think this may have accidentally snuck in on the last commit

SteveDMurphy commented on fidesctl/tests/core/test_generate_dataset.py:
```diff
@@ -617,3 +637,93 @@ def test_dataset_coverage_manifest_passes_snowflake(self, test_config, tmpdir):
+    }
+
+    @pytest.fixture(scope="class", autouse=True)
+    def redshift_setup(self):
```
More of a thought than anything to do for this here (likely feeding off of the great work reducing code duplication around `get_db_collections_and_fields`!) but it would be nice if we could refactor some of these tests to be more reusable as well instead of having duplication across tests per connector

ThomasLaPiana replied:
I think there is definitely some room for parameterization here, now that they all function much more similarly

@earmenda do you think a test refactor makes sense here or would you rather wait and see what happens with mongo?

earmenda replied:
I was a little stuck on this but it definitely makes sense to condense this. It might be simple to do this as a parameterized class but I still need to figure out now to make it work with the custom `external` mark i had added. I'm okay doing this with the mongo work to keep things moving along.

SteveDMurphy reviewed (changes requested):
Great work @earmenda (and @brentonmallen1) !! Everything looked quite good to me, it looks like there was possibly a generated `dataset.yml` that slipped in is all

ThomasLaPiana reviewed (approved):
LGTM!

earmenda commented:
@ThomasLaPiana @SteveDMurphy 

I refactored the test to used shared code now. I did have some challenges with having the external mark to exclude external databases so I changed the implementation to use a command line option instead. These tests are now skipped at runtime if you run pytest with `--ignore-external`

```
pytest --ignore-external
```

```
tests/core/test_generate_dataset.py::TestDatabase::test_get_db_tables[postgresql] PASSED                                                                                                  [ 72%]
tests/core/test_generate_dataset.py::TestDatabase::test_get_db_tables[mysql] PASSED                                                                                                       [ 72%]
tests/core/test_generate_dataset.py::TestDatabase::test_get_db_tables[mssql] PASSED                                                                                                       [ 72%]
tests/core/test_generate_dataset.py::TestDatabase::test_get_db_tables[snowflake] SKIPPED (Ignoring external database test as --ignore-external is set)                                    [ 73%]
tests/core/test_generate_dataset.py::TestDatabase::test_get_db_tables[redshift] SKIPPED (Ignoring external database test as --ignore-external is set)                                     [ 73%]
tests/core/test_generate_dataset.py::TestDatabase::test_generate_dataset[postgresql] PASSED                                                                                               [ 74%]
tests/core/test_generate_dataset.py::TestDatabase::test_generate_dataset[mysql] PASSED                                                                                                    [ 74%]
tests/core/test_generate_dataset.py::TestDatabase::test_generate_dataset[mssql] PASSED                                                                                                    [ 75%]
tests/core/test_generate_dataset.py::TestDatabase::test_generate_dataset[snowflake] SKIPPED (Ignoring external database test as --ignore-external is set)                                 [ 75%]
tests/core/test_generate_dataset.py::TestDatabase::test_generate_dataset[redshift] SKIPPED (Ignoring external database test as --ignore-external is set)                                  [ 76%]
tests/core/test_generate_dataset.py::TestDatabase::test_generate_dataset_passes_[postgresql] PASSED                                                                                       [ 76%]
tests/core/test_generate_dataset.py::TestDatabase::test_generate_dataset_passes_[mysql] PASSED                                                                                            [ 77%]
tests/core/test_generate_dataset.py::TestDatabase::test_generate_dataset_passes_[mssql] ERROR                                                                                             [ 77%]
tests/core/test_generate_dataset.py::TestDatabase::test_generate_dataset_passes_[snowflake] SKIPPED (Ignoring external database test as --ignore-external is set)                         [ 77%]
tests/core/test_generate_dataset.py::TestDatabase::test_generate_dataset_passes_[redshift] SKIPPED (Ignoring external database test as --ignore-external is set)                          [ 78%]
tests/core/test_generate_dataset.py::TestDatabase::test_generate_dataset_coverage_failure[postgresql] PASSED                                                                              [ 78%]
tests/core/test_generate_dataset.py::TestDatabase::test_generate_dataset_coverage_failure[mysql] PASSED                                                                                   [ 79%]
tests/core/test_generate_dataset.py::TestDatabase::test_generate_dataset_coverage_failure[mssql] PASSED                                                                                   [ 79%]
tests/core/test_generate_dataset.py::TestDatabase::test_generate_dataset_coverage_failure[snowflake] SKIPPED (Ignoring external database test as --ignore-external is set)                [ 80%]
tests/core/test_generate_dataset.py::TestDatabase::test_generate_dataset_coverage_failure[redshift] SKIPPED (Ignoring external database test as --ignore-external is set)                 [ 80%]
tests/core/test_generate_dataset.py::TestDatabase::test_dataset_coverage_manifest_passes[postgresql] PASSED                                                                               [ 81%]
tests/core/test_generate_dataset.py::TestDatabase::test_dataset_coverage_manifest_passes[mysql] PASSED                                                                                    [ 81%]
tests/core/test_generate_dataset.py::TestDatabase::test_dataset_coverage_manifest_passes[mssql] PASSED                                                                                    [ 81%]
tests/core/test_generate_dataset.py::TestDatabase::test_dataset_coverage_manifest_passes[snowflake] SKIPPED (Ignoring external database test as --ignore-external is set)                 [ 82%]
tests/core/test_generate_dataset.py::TestDatabase::test_dataset_coverage_manifest_passes[redshift] SKIPPED (Ignoring external database test as --ignore-external is set)                  [ 82%]
```

SteveDMurphy reviewed (approved):
Testing changes look great @earmenda ! I like the `--ignore-external` flag as well 👍🏽

ThomasLaPiana commented:
@earmenda merge when ready!
